From 23a3f7e5a23fd0ff043b1cb42607ba7272c1ce61 Mon Sep 17 00:00:00 2001
From: Dan McDonald <danmcd@joyent.com>
Date: Mon, 9 Apr 2018 18:07:37 -0400
Subject: [PATCH] OS-6872 mac deadlock in aggrs Reviewed by: Ryan Zezeski
 <rpz@joyent.com> Approved by: Ryan Zezeski <rpz@joyent.com>

---
 usr/src/uts/common/io/aggr/aggr_port.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/usr/src/uts/common/io/aggr/aggr_port.c b/usr/src/uts/common/io/aggr/aggr_port.c
index 74f34a3cd5..c8dbe00336 100644
--- a/usr/src/uts/common/io/aggr/aggr_port.c
+++ b/usr/src/uts/common/io/aggr/aggr_port.c
@@ -715,9 +715,10 @@ aggr_port_addvlan(aggr_port_t *port, uint_t idx, uint16_t vid)
 	 * nothing to do.
 	 */
 	if (port->lp_hwghs[idx] == NULL)
-		return (0);
+		err = 0;
+	else
+		err = mac_hwgroup_addvlan(port->lp_hwghs[idx], vid);
 
-	err = mac_hwgroup_addvlan(port->lp_hwghs[idx], vid);
 	mac_perim_exit(pmph);
 	return (err);
 }
@@ -737,9 +738,10 @@ aggr_port_remvlan(aggr_port_t *port, uint_t idx, uint16_t vid)
 		idx = 0;
 
 	if (port->lp_hwghs[idx] == NULL)
-		return (0);
+		err = 0;
+	else
+		err = mac_hwgroup_remvlan(port->lp_hwghs[idx], vid);
 
-	err = mac_hwgroup_remvlan(port->lp_hwghs[idx], vid);
 	mac_perim_exit(pmph);
 	return (err);
 }
-- 
2.21.0

